core/src/main/scala/org/apache/spark/rdd/ReliableCheckpointRDD.scala

```diff
@@ -279,8 +287,13 @@ private[spark] object ReliableCheckpointRDD extends Logging {
 
     // Register an on-task-completion callback to close the input stream.
     context.addTaskCompletionListener(context => deserializeStream.close())
-
-    deserializeStream.asIterator.asInstanceOf[Iterator[T]]
+    Utils.tryWithSafeFinally {
```

Comment:
This code will introduce issue, `deserializaStream` should be called after finished, the code here will close this stream prematurely.

Also please look at L289, it already takes care of `close` after the task is finished.